elements/lisk-tree/src/merkle_tree/utils.ts

```diff
@@ -142,3 +142,54 @@ export const getPairLocation = (nodeInfo: {
 
 	return pairLocation as NodeLocation;
 };
+
+export const calculateMerkleRoot = ({ value, appendPath, size }: MerkleRootInfo) => {
+	// 1. Calculate the new root
+
+	// Add prefix to value
+	const leafValueWithPrefix = Buffer.concat(
+		[LEAF_PREFIX, value],
+		LEAF_PREFIX.length + value.length,
+	);
+
+	// Set current hash to hash of value
+	const newLeafHash = hash(leafValueWithPrefix);
+	let currentHash = newLeafHash;
+
+	// Count the 1's in binaryLength
+	let count = 0;
+
+	const binaryLength = size.toString(2);
+
+	for (let i = 0; i < binaryLength.length; i += 1) {
+		// Loop the binaryLength from the right
+		// The right-most digits correspond to lower layers in the tree
+		if ((size >> i) & 1) {
+			const siblingHash = appendPath[count];
+			currentHash = generateHash(BRANCH_PREFIX, siblingHash, currentHash);
+			count += 1;
+		}
+	}
+
+	const newRoot = currentHash;
+
+	// 2. Update the append path
+	let subTreeIndex;
+	const treeHeight = Math.ceil(Math.log2(size)) + 1;
+
+	for (subTreeIndex = 0; subTreeIndex < treeHeight; subTreeIndex += 1) {
+		if (!((size >> subTreeIndex) & 1)) {
+			break;
+		}
+	}
+
+	let newAppendPath = appendPath.slice(0);
+	currentHash = newLeafHash;
+	const splicedPath = newAppendPath.splice(subTreeIndex);
```

Comment:
why are we using `splice` here instead of `splice`?

Reply:
It is to easily split the appendPath into two since `splice` mutates the original. We can also use slice but with splice it is a one-liner.